easybuild/tools/build_from_easystack.py

```diff
@@ -0,0 +1,173 @@
+
+
+class Easystack(object):
+    """One class instance per easystack. General options + list of all SoftwareSpecs instances"""
+    def __init__(self):
+        self.easybuild_version = None
+        self.robot = False
+        self.software_list = []
+
+    def compose_ec_names(self):
+        """Returns a list of all easyconfig names"""
+        ec_names = []
+        for sw in self.software_list:
+            ec_to_append = '%s-%s-%s-%s.eb' % (str(sw.software), str(sw.version),
+                                               str(sw.toolchain_name), str(sw.toolchain_version))
+            if ec_to_append is None:
+                continue
+            else:
+                ec_names.append(ec_to_append)
+        return ec_names
+
+    # flags applicable to all sw (i.e. robot)
+    def get_general_options(self):
+        """Returns general options (flags applicable to all sw (i.e. --robot))"""
+        general_options = {}
+        # TODO add support for general_options
+        # general_options['robot'] = self.robot
+        # general_options['easybuild_version'] = self.easybuild_version
+        return general_options
+
+
+class SoftwareSpecs(object):
+    """Contains information about every SW that should be installed"""
+    def __init__(self, software, version, toolchain, toolchain_version, toolchain_name):
+        self.software = software
+        self.version = version
+        self.toolchain = toolchain
+        self.toolchain_version = toolchain_version
+        self.toolchain_name = toolchain_name
+        self.toolchain = toolchain
+        self.versionsuffix = None
+
+    # to be implemented
+    # def get_versionsuffix(self):
+    #     return self.versionsuffix or ''
+
+
+# implement this to your own needs - to create custom yaml/json/xml parser
+class GenericSpecsParser(object):
+    """Parent of all implemented parser classes"""
+    @ staticmethod
+    def parse(filename):
+        """Parent of all implemented parser functions (i.e. JSON, XML...)"""
+        raise NotImplementedError
+
+
+class YamlSpecParser(GenericSpecsParser):
+    """YAML file parser"""
+    @ staticmethod
+    def parse(filename):
+        """Parses YAML file and assigns obtained values to SW config instances as well as general config instance"""
+        try:
+            with open(filename, 'r') as f:
+                spec_dict = yaml.safe_load(f)
+
+            eb = Easystack()
+        except Exception:
+            raise EasyBuildError("Could not read provided easystack.")
+
+        sw_dict = spec_dict["software"]
+
+        # assign software-specific EB attributes
+        for software in sw_dict:
+            try:
+                # iterates through toolchains to find out what sw version is needed
+                for yaml_toolchain in sw_dict[software]['toolchains']:
+                    # retrieves version number
+                    for yaml_version in sw_dict[software]['toolchains'][yaml_toolchain]['versions']:
```

Comment:
Please use a variable for stuff you're using in multiple places, and avoid using `yaml_` (at this point, it doesn't matter anymore where the data comes from?)

```python
for software_name, software_data in easystack_raw_dict.items():
    # make very sure we have a string value (parsed YAML data may have other value types)
    software_name = str(software_name)
    toolchains = software_data['toolchains']
    for toolchain_name, toolchain_data in toolchains.items():
        toolchain_name = str(toolchain_name)
        for key in toolchain_data['versions']:
            key = str(key)
```

I used `key` here in the inner loop since at this point you're not sure yet it's a version, if could be `'versionsuffix'` or `'exclude-labels'` (in the future).